drivers: rtc: Add support for Epson RX8130CE RTC driver

Mattemagikern opened the pull request:
This commit introduces a driver for the Epson RX8130CE rtc. The driver provides the following functionalities:

- Time setting and retrieval
- Periodic update interrupt support (1Hz)
- Alarm setting and retrieval (minute, hour, day)
- Frequency output control (32.768kHz, 1.024kHz, 1Hz, off)
- Power management (automatic power switching & battery charging)
- Calibration setting and retrieval

Mattemagikern committed:
drivers: rtc: Add support for Epson RX8130CE RTC driver

This commit introduces a driver for the Epson RX8130CE rtc.
The driver provides the following functionalities:

- Time setting and retrieval
- Periodic update interrupt support (1Hz)
- Alarm setting and retrieval (minute, hour, day)
- Frequency output control (32.768kHz, 1.024kHz, 1Hz, off)
- Power management (automatic power switching & battery charging)
- Calibration setting and retrieval

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

pdgendt commented on drivers/rtc/Kconfig.rx8130ce:
```diff
@@ -0,0 +1,11 @@
+	depends on DT_HAS_EPSON_RX8130CE_RTC_ENABLED
+	select I2C
+	help
+		Enable RTC driver for the Epson rx8130ce rtc
```
```suggestion
	  Enable RTC driver for the Epson rx8130ce rtc
```

pdgendt commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,758 @@
+ * 0x1E control0
+ * 0x1F ctrl1
+ */
+struct __attribute__((__packed__)) rx8130ce_registers {
```
Use the `__packed` macro.

pdgendt commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,758 @@
+	return 0;
+}
+
+static const struct rtc_driver_api rx8130ce_driver_api = {
```
Wrap using the API macros

```suggestion
static DEVICE_API(rtc, rx8130ce_driver_api) = {
```

pdgendt reviewed (dismissed):
Add a commit where you add the driver to `tests/drivers/build_all/rtc` so it is built in CI.

Mattemagikern committed:
tests: rtc: add epson rx8130ce to rtc build test

Include the Epson RX8130CE RTC driver in the RTC build tests to ensure it
compiles correctly and remains free of build regressions.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

bjarki-andreasen commented on dts/bindings/rtc/epson,rx8130ce.yaml:
```diff
@@ -0,0 +1,34 @@
+description: RX8130CE RTC
+
+compatible: "epson,rx8130ce-rtc"
+include: i2c-device.yaml
```
also include rtc-device.yaml which includes the `alarms-count` property :)

pdgendt commented on tests/drivers/build_all/rtc/i2c_devices.overlay:
```diff
@@ -84,6 +84,12 @@
 				reg = <0x7>;
 				irq-frequency = <512>;
 			};
+			test_rx8130ce: rx8130ce-rtc@32 {
```
Address needs to match `reg`

```suggestion
			test_rx8130ce: rx8130ce-rtc@8 {
```

pdgendt commented on tests/drivers/build_all/rtc/i2c_devices.overlay:
```diff
@@ -84,6 +84,12 @@
+				compatible = "epson,rx8130ce-rtc";
+				status = "okay";
+				reg = <0x8>;
+				irq-gpios = <&gpio0 0 0>;
```
```suggestion
				irq-gpios = <&test_gpio 0 0>;
```

Mattemagikern replied:
Thanks :)

bjarki-andreasen commented on dts/bindings/rtc/epson,rx8130ce.yaml:
```diff
@@ -0,0 +1,34 @@
+  irq-gpios:
+    type: phandle-array
+
+  clockout-frequency:
+    type: int
+    default: 0
+    description: |
+      Specify the ociillator frequency, the default is 0 (off)
+      - 0 # off
+      - 1 # 1 Hz oscillator
+      - 2 # 1 kHz oscillator
+      - 3 # 32 kHz oscillator
+    enum: [0, 1, 2, 3]
```
Providing the frequency here in Hz is easier to read than an enum
```
enum:
  - 1
  - 1000
  - 32768
```
to disable the output, just don't provide the property :) you can use DT_PROP_OR() from the driver :)

Mattemagikern replied:
Great tip, will fix!

bjarki-andreasen commented on dts/bindings/rtc/epson,rx8130ce.yaml:
```diff
@@ -0,0 +1,34 @@
+      - 3 # 32 kHz oscillator
+    enum: [0, 1, 2, 3]
+
+  battery-switchover:
+    type: int
+    default: 0
+    description: |
+      Enable the battery backup feature
+      the default is 0 (off)
+      - 0 # No battery backup
+      - 1 # Non-rechargeable battery backup i2c & FOUT are disabled when VDD is below Vdet1
+      - 2 # Non-rechargeable battery backup i2c & FOUT are always on
+      - 3 # Rechargeable battery backup i2c & FOUT are disabled when VDD is below Vdet1
+      - 4 # Rechargeable battery backup i2c & FOUT are always on
+    enum: [0, 1, 2, 3, 4]
```
If there exist readible names for these enums, using a string would possibly be easier to read:
```
type: string
enum:
  - "no-backup"
  - "non-recharge-backup-no-fout"
  ...
```

Mattemagikern replied:
The battery backup feature dosn't have readable names in the data sheet.
I'm going to remove the 0 entry, but I'd rather keep the rest.

bjarki-andreasen commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,757 @@
+};
+
+struct rx8130ce_data {
+	struct k_mutex lock;
```
Please avoid mutexes for locking, they are very expensive compared to a binary semaphore, and only required if thread ownership needs to be tracked:
```
k_sem_take();
k_sem_give();
```

Mattemagikern replied:
In `rx8130ce_irq_work_handler` the driver own the lock when the driver call the callback function(s) to the user. 

If I switch to a semaphore the callbacks would be unable to read time or set new alarms so I opt to keep it as is.

bjarki-andreasen replied:
Driver callbacks are usually not meant to be used to do work, most of them are called from interrupt contexts, including integrated RTCs. If you want the driver APIs to be callable from the callback, including setting or clearing the callback itself, its simpler to copy the callback and user data in `rx8130ce_irq_work_handler,` release the lock, then call the callback.

Mattemagikern replied:
Done :+1:

pdgendt commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,757 @@
+	return rc;
+
+}
+static int rx8130ce_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				  struct rtc_time *timeptr)
```
Nit

```suggestion
static int rx8130ce_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
				   struct rtc_time *timeptr)
```

pdgendt commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,757 @@
+	return rc;
+}
+
+static int rx8130ce_alarm_set_callback(const struct device *dev, uint16_t id,
+				      rtc_alarm_callback callback, void *user_data)
```
Nit

```suggestion
static int rx8130ce_alarm_set_callback(const struct device *dev, uint16_t id,
				       rtc_alarm_callback callback, void *user_data)
```

pdgendt commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,757 @@
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+static int rx8130ce_update_set_callback(const struct device *dev, rtc_update_callback callback,
+				       void *user_data)
```
Nit 

```suggestion
static int rx8130ce_update_set_callback(const struct device *dev, rtc_update_callback callback,
					void *user_data)
```

bjarki-andreasen reviewed (changes requested):
Have you tested the driver using https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/rtc/rtc_api ?

bjarki-andreasen commented:
Have you tested the driver using https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/rtc/rtc_api ?

Mattemagikern committed:
tests: rtc_api: Disable update callback after test completion

The update callback remains enabled after the test finishes, leaving the
RTC in a modified state. This change ensures the callback is properly
disabled, restoring the RTC to its original condition.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern commented:
@bjarki-andreasen 

Yes, I tested it using the rtc_api tests. Everything ran as expected (excep the y2k test but that has been ammended!)
It's great to see these tests included—they’ll be a valuable addition to our test suite for device validation!

```
TESTSUITE rtc_api succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [rtc_api]: pass = 7, fail = 0, skip = 0, total = 7 duration = 79.694 seconds
 - PASS - [rtc_api.test_alarm] duration = 26.033 seconds
 - PASS - [rtc_api.test_alarm_callback] duration = 26.048 seconds
 - PASS - [rtc_api.test_set_get_calibration] duration = 0.531 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.018 seconds
 - PASS - [rtc_api.test_time_counting] duration = 10.041 seconds
 - PASS - [rtc_api.test_update_callback] duration = 15.005 seconds
 - PASS - [rtc_api.test_y2k] duration = 2.018 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

bjarki-andreasen reviewed (dismissed):
Nice, great addition to the test suite as well!

Mattemagikern commented:
sorry @bjarki-andreasen and @pdgendt found a bug when I was writing some factory tests, I hope you would be so kind to re-rewiew the changes. 

Thank you for your patience ^^

bjarki-andreasen commented:
@Mattemagikern No need to apologize, you are contributing good work, and taking care to ensure a high quality of said work. That is commendable. Also, us reviewers don't sit around waiting for PRs to finalize, there is no rush :)

Mattemagikern commented on drivers/rtc/rtc_rx8130ce.c:
```diff
@@ -0,0 +1,776 @@
+					const struct rtc_time *timeptr)
+{
+	int rc = 0;
+	bool alarm_enabled;
```
I'll fix a PR for this and more in the weekend :+1:

Mattemagikern replied:
I'll add some test cases to the rtc_api to increase the coverage of the alarm :) 
#85247
